### PR TITLE
Implement upgrade in flight

### PR DESCRIPTION
**Pull request by imjfckm**
Implementing upgrade script for CAS.
Usage: `./upgrade start` to start upgrade in flight.

Whole mechanism is implemented using quite simple state machine. Sunny path is just progressing through sequence of steps, in case of errors it tries to unwind any changes made and get back to the state from before the upgrade. Also, in case no installation is detected there is an option to perform a fresh install.

### CODE REVIEW

File: utils/opencas.py [hunk 1]
```diff
@@ -154,9 +154,8 @@ def io_class_load_config(cls, cache_id, ioclass_file):
 
     @classmethod
     def start_upgrade(cls):
-        cmd = [cls.casadm_path,
-                    '--script',
-                    '--upgrade-in-flight']
+        cmd = [cls.casadm_path, '--script', '--upgrade-in-flight']
+
         return cls.run_cmd(cmd)
```
Comment (Ostrokrzew) [inline utils/opencas.py]: Shouldn't You merge this change with previous commit and move it there?

File: utils/upgrade [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+# SPDX-License-Identifier: BSD-3-Clause-Clear
+#
+
+import logging
+import sys
+import argparse
+import subprocess
+import os
+from pathlib import Path
+import re
```
Comment (Deixx) [inline utils/upgrade]: looks like following imports are not necessary: logging, subprocess, os, re (probably importing * from upgrade_utils makes them redundant)

File: utils/upgrade [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+        if not isinstance(result, Success):
+            return Failure("Dry run failed")
+
+        try:
+            logging.info(
+                "Insert with dry_run set succeeded. Removing module to attempt final insertion."
+            )
+            remove_module("cas_cache")
+        except Exception as e:
+            return Failure("Could not unload module. Reason: {e}")
```
Comment (Deixx) [inline utils/upgrade]: add f before string

File: utils/upgrade [hunk 4]
```diff
@@ -0,0 +1,345 @@
...
+            if current in ["noop", "none"]:
+                logging.info(f"Core {core} already uses {current} scheduler. Skipping.")
+            elif "noop" in available:
+                logging.info(f"Switching scheduler for {core}: {current} => noop")
+                set_device_scheduler(core, "noop")
+            elif "none" in available:
+                logging.info(f"Switching scheduler for {core}: {current} => none")
+                set_device_scheduler(core, "none")
+            else:
+                logging.info(f"No apropriate scheduler available for {core}. Skipping.")
```
Comment (Deixx) [inline utils/upgrade]: appropriate

File: utils/upgrade_utils.py [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+        if self.will_prompt:
+            print(f"{self.log+'...':60}", end="", flush=True)
+
+        print(self.result.result_mark())
+
+
+def insert_module(name, installed=True, **params):
+    cmd_params = [f"{param}={val}" for param, val in params.items()]
+
+    p = None
```
Comment (Deixx) [inline utils/upgrade_utils.py]: no need to initialize p as None

File: utils/upgrade_utils.py [hunk 6]
```diff
@@ -0,0 +1,209 @@
...
+    return subprocess.check_output(
+        f"find -L /sys/block -maxdepth 2 | grep {basename}$ |"
+        " sed -r 's/(\/sys\/block\/[^/]+).*/\\1/'",
+        shell=True,
+        encoding="ascii",
+    ).rstrip("\n")  # noqa W605
+
+
+def get_device_schedulers(sysfs_path):
+    schedulers = ""
```
Comment (Deixx) [inline utils/upgrade_utils.py]: no need to initialize schedulers

File: utils/upgrade [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+    log = "Setting core devices schedulers to noop"
+
+    def do_work(self):
+        unique_core_sysfs_devices = {
+            get_device_sysfs_path(core["device"])
+            for _, core in opencas.get_devices_state()["cores"].items()
+        }
+
+        self.state_machine.schedulers = {}
+        for core in unique_core_sysfs_devices:
```
Comment (Deixx) [inline utils/upgrade]: Maybe change core to core_sysfs_path? For me core is not self-explainable enough.

File: utils/upgrade [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+                logging.info(f"No apropriate scheduler available for {core}. Skipping.")
+
+        return Success()
+
+
+class RestoreCoreSchedulers(UpgradeState):
+    log = "Restoring core devices schedulers"
+
+    def do_work(self):
+        for core, scheduler in self.state_machine.schedulers.items():
```
Comment (Deixx) [inline utils/upgrade]: maybe change core to core_sysfs_path?

File: utils/upgrade [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+            current, _ = get_device_schedulers(core)
```
Comment (Deixx) [inline utils/upgrade]: I'd avoid '_' where possible - get_device_schedulers(core)[0] gets the job done

File: utils/upgrade [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+class InsertNewModule(InsertModule):
+    log = "Try to insert new caching module"
+
+
+class DryRun(InsertModule):
+    log = "Perform dry run to check upgrade data integrity"
+    options = {"installed": False, "dry_run": 1}
+
+    def do_work(self):
+        result = super().do_work()
```
Comment (Deixx) [inline utils/upgrade]: Wouldn't this be the place to workaround checking available/free memory?

File: utils/upgrade (hunk 2 above)
Comment (imjfckm) [reply utils/upgrade]: Changed it to not use wildcard import

File: utils/upgrade (hunk 10 above)
Comment (imjfckm) [reply utils/upgrade]: Included workaround in a separate state

File: utils/opencas.py (hunk 1 above)
Comment (imjfckm) [reply utils/opencas.py]: Done

File: utils/upgrade [hunk 11]
```diff
@@ -0,0 +1,345 @@
...
+        return Success()
+
+
+class BuildCas(UpgradeState):
+    log = "Compiling Open CAS"
+
+    def do_work(self):
+        with open(COMPILATION_LOG, "w") as build_log:
+            logging.info("Running ./configure for CAS")
+            p = subprocess.run(["./configure"], cwd="..", stdout=build_log, stderr=build_log)
```
[the remark this answers is not in the document]
Comment (imjfckm) [reply utils/upgrade]: Done

File: utils/upgrade [hunk 12]
```diff
@@ -0,0 +1,345 @@
...
+    log = "Restoring core devices schedulers"
+
+    def do_work(self):
+        for core, scheduler in self.state_machine.schedulers.items():
+            current, _ = get_device_schedulers(core)
+            if scheduler in ["noop", "none"]:
+                logging.info(f"Device {core} already uses {scheduler} scheduler. Skipping.")
+            else:
+                logging.info(f"Switching scheduler for {core}: {current} => {scheduler}")
+            set_device_scheduler(core, scheduler)
```
[the remark this answers is not in the document]
Comment (imjfckm) [reply utils/upgrade]: Yup, corrected

File: utils/upgrade [hunk 13]
```diff
@@ -0,0 +1,345 @@
...
+    parser = argparse.ArgumentParser(prog=__file__)
+    subparsers = parser.add_subparsers(title="actions")
+
+    parser_start = subparsers.add_parser(
+        "start", help="Upgrade Open CAS in flight to current version"
+    )
+    parser_start.add_argument(
+        "--force",
+        action="store_true",
+        help="Discard any leftover upgrade state and start from beginning",
```
[the remark this answers is not in the document]
Comment (imjfckm) [reply utils/upgrade]: Yup!

File: utils/upgrade [hunk 14]
```diff
@@ -0,0 +1,345 @@
...
+            return self.NotInstalled(f"Couldn't detect current CAS version")
+        except Exception as e:
+            return Failure(f"Failed to get current version of CAS {e}")
+
+        if version[CAS_CLI_KEY] != version[CAS_CACHE_KEY]:
+            return Failure("Mismatch between CLI and cas_cache version")
+
+        active_devices = opencas.get_devices_state()
+
+        if any([device["status"] != "Running" for _, device in active_devices["caches"].items()]):
```
[the remark this answers is not in the document]
Comment (imjfckm) [reply utils/upgrade]: Done

File: utils/upgrade [hunk 15]
```diff
@@ -0,0 +1,345 @@
...
+        return Success()
+
+
+class SetSchedulersToNoop(UpgradeState):
+    log = "Setting core devices schedulers to noop"
+
+    def do_work(self):
+        unique_core_sysfs_devices = {
+            get_device_sysfs_path(core["device"])
+            for _, core in opencas.get_devices_state()["cores"].items()
```
[the remark this answers is not in the document]
Comment (imjfckm) [reply utils/upgrade]: Fixed

File: utils/upgrade (hunk 3 above)
Comment (imjfckm) [reply utils/upgrade]: Done

File: utils/upgrade_utils.py (hunk 6 above)
Comment (imjfckm) [reply utils/upgrade_utils.py]: Yup, fixed

File: utils/upgrade (hunk 7 above)
Comment (imjfckm) [reply utils/upgrade]: Done

File: utils/upgrade (hunk 8 above)
Comment (imjfckm) [reply utils/upgrade]: Ditto

File: utils/upgrade (hunk 9 above)
Comment (imjfckm) [reply utils/upgrade]: Fixed

File: utils/upgrade_utils.py (hunk 5 above)
Comment (imjfckm) [reply utils/upgrade_utils.py]: Fixed

File: utils/upgrade (hunk 4 above)
Comment (imjfckm) [reply utils/upgrade]: Done

**Pull request merged**